{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b736165f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from collections import defaultdict\n",
    "from Bio.Align import AlignInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0728b2fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_temp_fasta(records, filepath):\n",
    "    \"\"\"\n",
    "    Write a list of SeqRecord objects to a FASTA file.\n",
    "\n",
    "    Parameters:\n",
    "    records (list of SeqRecord): List of sequence records to write to the file.\n",
    "    filepath (str): Path to the output FASTA file.\n",
    "    \"\"\"\n",
    "    SeqIO.write(records, filepath, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5f11eada",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_sequences_by_class(input_fasta):\n",
    "    \"\"\"\n",
    "    Parse sequences from a FASTA file and group them by class.\n",
    "\n",
    "    Assumes that the class label is the last item in the description, separated by '|'.\n",
    "\n",
    "    Parameters:\n",
    "    input_fasta (str): Path to the input FASTA file containing sequences.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary mapping class labels to lists of SeqRecord objects.\n",
    "    \"\"\"\n",
    "    sequences_by_class = defaultdict(list)\n",
    "    for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "        sequence_class = record.description.split('|')[-1]\n",
    "        sequences_by_class[sequence_class].append(record)\n",
    "    return sequences_by_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4746dd44",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_sequences(mafft_path, input_fasta, output_fasta):\n",
    "    \"\"\"\n",
    "    Perform multiple sequence alignment using MAFFT.\n",
    "\n",
    "    Parameters:\n",
    "    mafft_path (str): Absolute path to the MAFFT executable.\n",
    "    input_fasta (str): Path to the input FASTA file containing unaligned sequences.\n",
    "    output_fasta (str): Path to the output FASTA file to save aligned sequences.\n",
    "\n",
    "    Notes:\n",
    "    - Uses '--maxiterate' and '--thread' options for MAFFT to improve alignment.\n",
    "    - Suppresses MAFFT output with '--quiet' for cleaner logs.\n",
    "    \"\"\"\n",
    "    command = [\n",
    "        mafft_path,\n",
    "        '--maxiterate', '1000',\n",
    "        '--thread', '20',\n",
    "        '--quiet',\n",
    "        input_fasta\n",
    "    ]\n",
    "    print(f\"Executing command: {' '.join(command)}\")\n",
    "\n",
    "    with open(output_fasta, 'w') as outfile:\n",
    "        try:\n",
    "            subprocess.run(\n",
    "                command,\n",
    "                check=True,\n",
    "                stdout=outfile,\n",
    "                stderr=subprocess.PIPE,\n",
    "                text=True\n",
    "            )\n",
    "            print(f\"MAFFT successfully executed for {input_fasta}\")\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"Error executing MAFFT for {input_fasta}: {e.stderr}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2c19cd9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_alignment(aligned_fasta):\n",
    "    \"\"\"\n",
    "    Read an aligned FASTA file and return a MultipleSeqAlignment object.\n",
    "\n",
    "    Parameters:\n",
    "    aligned_fasta (str): Path to the aligned FASTA file.\n",
    "\n",
    "    Returns:\n",
    "    MultipleSeqAlignment: An object containing aligned sequences.\n",
    "    \"\"\"\n",
    "    return AlignIO.read(aligned_fasta, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "87724513",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_consensus(align):\n",
    "    \"\"\"\n",
    "    Generate a consensus sequence from a multiple sequence alignment.\n",
    "\n",
    "    Parameters:\n",
    "    align (MultipleSeqAlignment): Alignment object containing aligned sequences.\n",
    "\n",
    "    Returns:\n",
    "    Seq: Consensus sequence as a Seq object.\n",
    "    \"\"\"\n",
    "    summary_align = AlignInfo.SummaryInfo(align)\n",
    "    consensus_seq = summary_align.gap_consensus(threshold=0.7, ambiguous='N')\n",
    "    return consensus_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ede277a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_consensus_per_class(input_fasta, output_fasta, mafft_path):\n",
    "    \"\"\"\n",
    "    Generate consensus sequences for each class in the input FASTA file.\n",
    "\n",
    "    Parameters:\n",
    "    input_fasta (str): Path to the input FASTA file containing sequences with class labels.\n",
    "    output_fasta (str): Path to the output FASTA file to save consensus sequences.\n",
    "    mafft_path (str): Absolute path to the MAFFT executable.\n",
    "\n",
    "    Workflow:\n",
    "    - Processes sequences grouped by class.\n",
    "    - Performs multiple sequence alignment using MAFFT.\n",
    "    - Generates a consensus sequence for each class.\n",
    "    - Writes the consensus sequences to the output FASTA file.\n",
    "    \"\"\"\n",
    "    sequences_by_class = process_sequences_by_class(input_fasta)\n",
    "    consensus_records = []\n",
    "\n",
    "    for cls, records in sequences_by_class.items():\n",
    "        temp_fasta = f'temp_{cls}.fasta'\n",
    "        aligned_fasta = f'{cls}_aligned.fasta'\n",
    "        \n",
    "        if not os.path.exists(aligned_fasta):\n",
    "            write_temp_fasta(records, temp_fasta)\n",
    "            align_sequences(mafft_path, temp_fasta, aligned_fasta)\n",
    "            if os.path.exists(temp_fasta):\n",
    "                os.remove(temp_fasta)\n",
    "\n",
    "        if os.path.exists(aligned_fasta):\n",
    "            align = read_alignment(aligned_fasta)\n",
    "            os.remove(aligned_fasta)\n",
    "        else:\n",
    "            print(f\"Failed to generate alignment for class {cls}\")\n",
    "            continue\n",
    "\n",
    "        consensus_seq = generate_consensus(align)\n",
    "        consensus_records.append(SeqRecord(consensus_seq, id=cls, description=\"\"))\n",
    "\n",
    "    SeqIO.write(consensus_records, output_fasta, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a2cce81d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing ../data/Human_betaherpesvirus_5/US28/US28_consensus.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_A1.fasta\n",
      "MAFFT successfully executed for temp_A1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_A2.fasta\n",
      "MAFFT successfully executed for temp_A2.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_D.fasta\n",
      "MAFFT successfully executed for temp_D.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B2.fasta\n",
      "MAFFT successfully executed for temp_B2.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B1.fasta\n",
      "MAFFT successfully executed for temp_B1.fasta\n",
      "Processing ../data/Human_betaherpesvirus_5/UL55/UL55_consensus.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_1.fasta\n",
      "MAFFT successfully executed for temp_1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_2.fasta\n",
      "MAFFT successfully executed for temp_2.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_4.fasta\n",
      "MAFFT successfully executed for temp_4.fasta\n",
      "Processing ../data/Human_betaherpesvirus_5/UL73/UL73_consensus.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_4b.fasta\n",
      "MAFFT successfully executed for temp_4b.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_4c.fasta\n",
      "MAFFT successfully executed for temp_4c.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_1.fasta\n",
      "MAFFT successfully executed for temp_1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_4a.fasta\n",
      "MAFFT successfully executed for temp_4a.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_3b.fasta\n",
      "MAFFT successfully executed for temp_3b.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_3a.fasta\n",
      "MAFFT successfully executed for temp_3a.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_2.fasta\n",
      "MAFFT successfully executed for temp_2.fasta\n",
      "Processing ../data/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_consensus.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.1.7.fasta\n",
      "MAFFT successfully executed for temp_B.1.1.7.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_JN.1.fasta\n",
      "MAFFT successfully executed for temp_JN.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_JN.1.11.1.fasta\n",
      "MAFFT successfully executed for temp_JN.1.11.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.1.fasta\n",
      "MAFFT successfully executed for temp_BA.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.2.12.1.fasta\n",
      "MAFFT successfully executed for temp_BA.2.12.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.2.fasta\n",
      "MAFFT successfully executed for temp_B.1.2.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.2.fasta\n",
      "MAFFT successfully executed for temp_BA.2.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.2.86.1.fasta\n",
      "MAFFT successfully executed for temp_BA.2.86.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_XBB.2.3.fasta\n",
      "MAFFT successfully executed for temp_XBB.2.3.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_XBB.fasta\n",
      "MAFFT successfully executed for temp_XBB.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.1.fasta\n",
      "MAFFT successfully executed for temp_B.1.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_CH.1.1.fasta\n",
      "MAFFT successfully executed for temp_CH.1.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_XBB.1.5.fasta\n",
      "MAFFT successfully executed for temp_XBB.1.5.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.fasta\n",
      "MAFFT successfully executed for temp_B.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_EG.5.1.fasta\n",
      "MAFFT successfully executed for temp_EG.5.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.617.2.fasta\n",
      "MAFFT successfully executed for temp_B.1.617.2.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.526.fasta\n",
      "MAFFT successfully executed for temp_B.1.526.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.429.fasta\n",
      "MAFFT successfully executed for temp_B.1.429.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.4.fasta\n",
      "MAFFT successfully executed for temp_BA.4.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BQ.1.fasta\n",
      "MAFFT successfully executed for temp_BQ.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_C.37.fasta\n",
      "MAFFT successfully executed for temp_C.37.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_P.1.fasta\n",
      "MAFFT successfully executed for temp_P.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.621.fasta\n",
      "MAFFT successfully executed for temp_B.1.621.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.427.fasta\n",
      "MAFFT successfully executed for temp_B.1.427.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.177.fasta\n",
      "MAFFT successfully executed for temp_B.1.177.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.525.fasta\n",
      "MAFFT successfully executed for temp_B.1.525.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.351.fasta\n",
      "MAFFT successfully executed for temp_B.1.351.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.5.fasta\n",
      "MAFFT successfully executed for temp_BA.5.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.1.1.fasta\n",
      "MAFFT successfully executed for temp_B.1.1.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.617.1.fasta\n",
      "MAFFT successfully executed for temp_B.1.617.1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.1.1.529.fasta\n",
      "MAFFT successfully executed for temp_B.1.1.529.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_XBB.1.16.fasta\n",
      "MAFFT successfully executed for temp_XBB.1.16.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_BA.2.75.fasta\n",
      "MAFFT successfully executed for temp_BA.2.75.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_D.2.fasta\n",
      "MAFFT successfully executed for temp_D.2.fasta\n",
      "Processing ../data/Human_immunodeficiency_virus_1/Human_immunodeficiency_virus_1_consensus.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_02_AG.fasta\n",
      "MAFFT successfully executed for temp_02_AG.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_A1.fasta\n",
      "MAFFT successfully executed for temp_A1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_01_AE.fasta\n",
      "MAFFT successfully executed for temp_01_AE.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_B.fasta\n",
      "MAFFT successfully executed for temp_B.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_G.fasta\n",
      "MAFFT successfully executed for temp_G.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_C.fasta\n",
      "MAFFT successfully executed for temp_C.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_D.fasta\n",
      "MAFFT successfully executed for temp_D.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_F1.fasta\n",
      "MAFFT successfully executed for temp_F1.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_O.fasta\n",
      "MAFFT successfully executed for temp_O.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_35_A1D.fasta\n",
      "MAFFT successfully executed for temp_35_A1D.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_08_BC.fasta\n",
      "MAFFT successfully executed for temp_08_BC.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_A6.fasta\n",
      "MAFFT successfully executed for temp_A6.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_11_cpx.fasta\n",
      "MAFFT successfully executed for temp_11_cpx.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_20_BG.fasta\n",
      "MAFFT successfully executed for temp_20_BG.fasta\n",
      "Executing command: c:\\Users\\lebat\\Documents\\KSS\\lib\\mafft\\mafft.bat --maxiterate 1000 --thread 20 --quiet temp_63_02A6.fasta\n",
      "MAFFT successfully executed for temp_63_02A6.fasta\n"
     ]
    }
   ],
   "source": [
    "# List of input and output file paths\n",
    "files = [\n",
    "    ('../data/Human_betaherpesvirus_5/US28/US28_nucleotide_sequences.fasta',\n",
    "     '../data/Human_betaherpesvirus_5/US28/US28_consensus.fasta'),\n",
    "    ('../data/Human_betaherpesvirus_5/UL55/UL55_nucleotide_sequences.fasta',\n",
    "     '../data/Human_betaherpesvirus_5/UL55/UL55_consensus.fasta'),\n",
    "    ('../data/Human_betaherpesvirus_5/UL73/UL73_nucleotide_sequences.fasta',\n",
    "     '../data/Human_betaherpesvirus_5/UL73/UL73_consensus.fasta'),\n",
    "    ('../data/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_nucleotide_sequences.fasta',\n",
    "     '../data/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_consensus.fasta'),\n",
    "    ('../data/Human_immunodeficiency_virus_1/Human_immunodeficiency_virus_1_nucleotide_sequences.fasta',\n",
    "     '../data/Human_immunodeficiency_virus_1/Human_immunodeficiency_virus_1_consensus.fasta')\n",
    "]\n",
    "\n",
    "# Absolute path to the MAFFT executable\n",
    "mafft_path = os.path.abspath(\"../lib/mafft/mafft.bat\")\n",
    "\n",
    "# Check if MAFFT executable exists before proceeding\n",
    "if not os.path.exists(mafft_path):\n",
    "    print(f\"MAFFT executable not found at {mafft_path}\")\n",
    "else:\n",
    "    # Process each pair of input and output files\n",
    "    for input_fasta, output_fasta in files:\n",
    "        print(f\"Processing {output_fasta}\")\n",
    "        generate_consensus_per_class(input_fasta, output_fasta, mafft_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
